phoenix-core/src/main/java/org/apache/phoenix/mapreduce/VerifyReplicationTool.java

```diff
@@ -0,0 +1,477 @@
+    static final Option
+            TARGET_ZK_QUORUM_OPT =
+            new Option("tz", "target-zookeeper", true,
+                    "Target ZooKeeper connection details (optional)");
+    static final Option
+            CONDITIONS_OPT =
+            new Option("c", "conditions", true,
+                    "Conditions for select query WHERE clause (optional)");
+    static final Option TIMESTAMP =
+            new Option("ts", "timestamp", true,
+                    "Timestamp in millis used to compare the two tables.  Defaults to current time minus 60 seconds");
+
+    static final Option HELP_OPT = new Option("h", "help", false, "Show this help and quit");
+
+    private Configuration conf;
+
+    private String zkQuorum;
+    private String tableName;
+    private String targetTableName;
+    private String targetZkQuorum;
+    private String sqlConditions;
+    private long timestamp;
+
+    VerifyReplicationTool(Configuration conf) {
+        this.conf = Preconditions.checkNotNull(conf, "Configuration cannot be null");
+    }
+
+    public static Builder newBuilder(Configuration conf) {
+        return new Builder(conf);
+    }
+
+    public static class Verifier
+            extends Mapper<NullWritable, VerifyReplicationSourceWritable, NullWritable, NullWritable> {
+
+        private QueryPlan targetQueryPlan;
+        private PhoenixResultSet targetResultSet = null;
+        private boolean targetHasNext;
+        private boolean sourceHasData;
+
+        public enum Counter {
+            GOODROWS, BADROWS, ONLY_IN_SOURCE_TABLE_ROWS, ONLY_IN_TARGET_TABLE_ROWS,
+            CONTENT_DIFFERENT_ROWS
+        }
+
+        @Override
+        protected void setup(Context context) throws IOException, InterruptedException {
+            super.setup(context);
+            try {
+                targetQueryPlan = PhoenixMapReduceUtil.getQueryPlan(context.getConfiguration(), true);
+            } catch (SQLException e) {
+                throw new IOException(e.getMessage());
+            }
+
+        }
+
+        @Override
+        protected void map(NullWritable key, VerifyReplicationSourceWritable value, Context context)
+                throws IOException, InterruptedException {
+
+            sourceHasData = true;
+            try {
+                if(targetResultSet == null) {
+                    Configuration conf = context.getConfiguration();
+                    byte[] targetStartRow = null;
+                    byte[] targetStopRow = null;
+                    // find source table split
+                    PhoenixInputSplit sourceInputSplit = (PhoenixInputSplit) context.getInputSplit();
+                    if(key != null) {
+                        targetStartRow = value.getSourceKey().get();
+                    }
+                    if(sourceInputSplit.getLength() != 0) {
+                        targetStopRow = sourceInputSplit.getKeyRange().getUpperRange();
```

Comment:
I am somewhat doubtful about setting the target start row using upper key range function. @twdsilva Do you see any problem with this?

Reply:
The regionSize or Length variable can be 0 because certain config props may disable the regionSizeCalculator, in which case it will always return 0. I would advise to check the `getUpperRange()` to see if there is end to the range and assign it appropriately.

Reply:
@karanmehta93  it always uses RegionSizeCalculator to generate the splits right? I don't see a problem with setting the targetStopRow based on the upper range of the split. Will the key ever be null? If it is, should the targetStartRow be set to the lower range of the split?

Reply:
@karanmehta93 one of the test covers the region size = 0 scenario, it seems to be working fine. I will check more. 
@twdsilva The key cannot be null, if the source splits doesn't have results it never enter the map task based on the record reader logic.

Reply:
> one of the test covers the region size = 0 scenario, it seems to be working fine. I will check more

I think all tests are getting region size as 0.